models/helper.go

```diff
@@ -21,47 +21,36 @@ func QueryFilterByConceptDefsPlusCohortPairsHelper(query *gorm.DB, sourceId int,
 	// Caching of temporary tables: for optimal performance, a temporary table dictionary / cache is updated, keeping a mapping
 	// of existing temporary table names vs underlying subsets of items in filterConceptDefsAndCohortPairs that gave rise to these
 	// tables.
+	// Returns: the final query,
+	// and, if the last variable in filterConceptDefsAndCohortPairs is a concept/numeric one,
+	// this method also returns the final observation table alias representing the filter
+	// added for this variable to the final query.
 	query = query.Table(resultsDataSource.Schema+".cohort as "+finalCohortAlias).
 		Select("*").
-		Where("cohort_definition_id=?", mainCohortDefinitionId)
+		Where(finalCohortAlias+".cohort_definition_id=?", mainCohortDefinitionId)
 
-	tmpTableName := ""
+	finalObservationTableAlias := ""
 	var err error
 	for i, item := range filterConceptDefsAndCohortPairs {
 		observationTableAlias := fmt.Sprintf("filter_%d", i)
 		switch convertedItem := item.(type) {
 		case utils.CustomConceptVariableDef:
-			query, tmpTableName, err = QueryFilterByConceptDefHelper(query, sourceId, convertedItem, omopDataSource, finalCohortAlias, observationTableAlias)
+			query, finalObservationTableAlias, err = QueryFilterByConceptDefHelper(query, sourceId, convertedItem, omopDataSource, finalCohortAlias, observationTableAlias)
 		case utils.CustomDichotomousVariableDef:
 			query = QueryFilterByCohortPairHelper(query, convertedItem, resultsDataSource, mainCohortDefinitionId, finalCohortAlias+".subject_id", observationTableAlias)
-			tmpTableName = ""
+			finalObservationTableAlias = ""
```

Comment:
I may be missing something but it seems like we do not need to do this again (same as line 32)

Reply:
we do need it. Line 41 is in a loop. The goal is to have this variable be empty if the last item is of type `CustomDichotomousVariableDef` and filled with the last temp table otherwise.
Having said that, I do realise that this implementation is a bit convoluted here...and there should be a better way than have `QueryFilterByConceptDefsPlusCohortPairsHelper` return different things in different scenarios.... Looking at the code again, I think it only serves the purpose of (maybe) improving the performance of `RetrieveHistogramDataBySourceIdAndCohortIdAndConceptDefsPlusCohortPairs`. I can make a new ticket to simplify this.